### ISSUE TITLE
repetition_penalty

### ISSUE THREAD

**Issue by edwardcapriolo** [edwardcapriolo/deliverance#20]
https://www.reddit.com/r/LocalLLaMA/comments/1g383mq/repetition_penalties_are_terribly_implemented_a/

 Part 1 - Frequency/presence/repetition penalty

Frequency and presence penalties are subtractive. Frequency penalty reduces word weights per existing word instance, whereas presence penalty reduces based on boolean word existence. Note that these penalties are applied to the logits (unnormalised weight predictions) of each token, not the final probability.

final_logit["word"] -> raw_logit["word"] - 
                       (word_count["word"] * frequency_penalty) -
                       (min(word_count["word"], 1) * presence_penalty)

Repetition penalty is the same as presence penalty, but multiplicative. This is usually good when trying different models, since the raw logit magnitude differs between models.

final_logit["word"] -> raw_logit["word"] / repetition_penalty^min(word_count["word"], 1)

A big problem for small models. something here should help.
